controllers/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,401 @@
+			return ctrl.Result{}, err
+		}
+
+		// Update status
+		instance.Status.ACID = newID
+		instance.Status.SecretName = secretName
+		instance.Status.CreatedAt = &metav1.Time{Time: time.Now().UTC()}
+		instance.Status.ExpiresAt = &metav1.Time{Time: expiresAt}
+
+		instance.Status.Conditions.MarkTrue(condition.ReadyCondition, "AC is ready")
+
+		// Immediate patch
+		if patchErr := helperObj.PatchInstance(ctx, instance); patchErr != nil {
+			return ctrl.Result{}, patchErr
+		}
+
+		// Requeue
+		nextCheck := r.computeNextRequeue(instance)
+		logger.Info("Rotated AC, requeuing to check again", "nextCheck", nextCheck)
+		return ctrl.Result{RequeueAfter: nextCheck}, nil
+	}
+
+	logger.Info("AC is up to date", "ACID", instance.Status.ACID)
+	instance.Status.Conditions.MarkTrue(condition.ReadyCondition, "Up to date")
+
+	// Requeue to check again before next expiry
+	nextCheck := r.computeNextRequeue(instance)
+	return ctrl.Result{RequeueAfter: nextCheck}, nil
+}
+
+func (r *ApplicationCredentialReconciler) reconcileDelete(
+	ctx context.Context,
+	instance *keystonev1.ApplicationCredential,
+	helperObj *helper.Helper,
+) (ctrl.Result, error) {
+
+	logger := r.GetLogger(ctx)
+
+	// Only if user explicitly does "oc delete" do we revoke the AC
+	if instance.Status.ACID != "" {
+		keystoneAPI, err := keystonev1.GetKeystoneAPI(ctx, helperObj, instance.Namespace, nil)
+		if err == nil && keystoneAPI.IsReady() {
+			userID, userErr := r.findUserIDAsAdmin(ctx, helperObj, keystoneAPI, instance.Spec.UserName)
+			if userErr == nil {
+				userOS, userRes, userErr2 := keystonev1.GetUserServiceClient(ctx, helperObj, keystoneAPI, instance.Spec.UserName)
+				if userErr2 == nil && userRes == (ctrl.Result{}) && userOS != nil {
+					delErr := applicationcredentials.Delete(userOS.GetOSClient(), userID, instance.Status.ACID).ExtractErr()
+					if delErr != nil {
+						logger.Error(delErr, "Failed to delete AC from Keystone", "ACID", instance.Status.ACID)
+					} else {
+						logger.Info("Deleted AC from Keystone", "ACID", instance.Status.ACID)
+					}
+				}
+			}
+		}
+	}
+
+	// Remove finalizer
+	const finalizer = "openstack.org/applicationcredential"
+	if controllerutil.RemoveFinalizer(instance, finalizer) {
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// createACWithName creates a new AC in Keystone
+func (r *ApplicationCredentialReconciler) createACWithName(
+	logger logr.Logger,
+	identClient *gophercloud.ServiceClient,
+	userID string,
+	ac *keystonev1.ApplicationCredential,
+	newACName string,
+) (string, string, time.Time, error) {
+	createOpts := applicationcredentials.CreateOpts{
+		Name:         newACName,
+		Description:  fmt.Sprintf("Created by keystone-operator for user %s", ac.Spec.UserName),
+		Unrestricted: false,
+
+		// Always assign these roles:
+		Roles: []applicationcredentials.Role{
+			{Name: "admin"},
+			{Name: "service"},
+		},
```

Comment:
This makes this controller less generic and therefore potentially less useful.   Perhaps this is another parameter - like the access rules that should be passed in as part of the AC spec.

The same can also be said for the Unrestricted field.

Reply:
I see there is support for all of Roles, Unrestricted and AccessRules in the gophercloud call - https://pkg.go.dev/github.com/gophercloud/gophercloud/openstack/identity/v3/applicationcredentials#CreateOpts